### DISCUSSION TITLE
Lesson11: Base64.encode(abi.encodePacked(svg)) == Base64.encode(bytes(string(abi.encodePacked(svg)))), is that possible?

### DISCUSSION THREAD

**Discussion by codeplay22** [Cyfrin/foundry-full-course-cu#1996] [Q&A] [answered]
in DeployMoodNft.s.sol  
in this code ``` Base64.encode(bytes(string(abi.encodePacked(svg)))) ``` why we are converting it to bytes and then to ```strings``` and then again to ```bytes``` and then again to ```strings(Base64)```. if we simply can do this ```Base64.encode(abi.encodePacked(svg)``` which gives the same output as well. check out in the image below.

![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/116997721/57f6523d-bca0-41bb-acf9-4000eb5d2203)

![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/116997721/c3925e97-aa67-498d-9137-46b0b8c4eaac)

**Comment by EngrPips** [comment]
Hello @codeplay22 That is a good catch, and I must say you are correct, but Patrick is a really security-minded person, and he will always do everything to be safe rather than sorry. Some things seem to work the same on a surface level, but when they are battle-tested, you will get to realize why the extra step was needed. This doesn't mean that doing it the way you have stated will definitely lead to a security issue, but I am just talking from the perspective with which I look at it when I see Patrick  do stuff in more complicated ways

**Reply by codeplay22** [reply]
@EngrPips yeah, Thanks for taking your time looking at my code brother, but i want to know exactly why he did that way?

**Comment by codeplay22** [comment]
Hi @PatrickAlphaC  i wanted to know why you first encoded it from ```bytes to strings then again to bytes``` and then finally to ```strings``` by using this``` Base64.encode(bytes(string(abi.encodePacked(svg))));``` ? whereas you could just do ``` Base64.encode(abi.encodePacked(svg))``` instead.
why did you do that?

**Reply by EngrPips** [reply]
I am looking forward to Patrick's response myself.

**Reply by PatrickAlphaC** [reply]
Whoops! Probably because I wasn't thinking clearly!

**Reply by codeplay22** [reply]
Hi @PatrickAlphaC 
Thanks a lot for taking time to read my code, i understand it happens.
Thanks a lot for replying though.
But i was wondering what to mark the answer?

**Reply by EngrPips** [reply]
> Whoops! Probably because I wasn't thinking clearly!

This is the answer because you wanted to why and clearly here he stated why

**Reply by codeplay22** [reply, marked as answer]
@EngrPips no, i asked that to know the concept behind it only.
So according to patrick i think answer would be to use it like this only ``` Base64.encode(abi.encodePacked(svg)) ```

and the same thing can be done here as well in ``` Moodnft.s.sol```
![image](https://github.com/user-attachments/assets/fb2fa6e1-9586-4072-834e-e89c719b9b7d)

there's no need to convert ```abi.EncodePacked to bytes``` you can remove it and it will work fine.

Like that -
![image](https://github.com/user-attachments/assets/4b627426-4875-4923-b6d5-8d6c451fea4b)

**Reply by EngrPips** [reply]
I see.

**Reply by programmer-ke** [reply]
Good to find this thread, I was wondering so myself.

**Reply by programmer-ke** [reply]
> Hi @PatrickAlphaC i wanted to know why you first encoded it from `bytes to strings then again to bytes` and then finally to `strings` by using this` Base64.encode(bytes(string(abi.encodePacked(svg))));` ? whereas you could just do ` Base64.encode(abi.encodePacked(svg))` instead. why did you do that?

I also wonder if there's much of a difference between

`Base64.encode(abi.encodePacked(svg))` and `Base64.encode(bytes(svg))`.

According to this [post][zep] one is copying memory and the other is casting a pointer type. It's from a while ago so things may have changed.

[zep]: https://forum.openzeppelin.com/t/difference-between-abi-encodepacked-string-and-bytes-string/11837/2

**Reply by EngrPips** [reply]
The difference boils down to what happens under the hood, which yields the same result.

**Reply by programmer-ke** [reply]
The next set of lessons on advanced EVM literally addressed this. All good!

**Reply by EngrPips** [reply]
> The next set of lessons on advanced EVM literally addressed this. All good!

Interesting
